### ISSUE TITLE
Possible unintended reference comparison

### ISSUE THREAD

**Issue by tvsonar** [SonarSource-VisualStudio/sonarlint-roslyn-analyzer#347]
R# ID: _PossibleUnintendedReferenceComparison_

C# is different from Java in this matter, so a simple merge into https://jira.sonarsource.com/browse/RSPEC-1698 won't work here. We'll need to create a subtask for C#. 

To compare objects in C# you can use `==` or `.Equals()`. For classes, you can overload the `operator ==`, so you may do value comparison with `==` as well. `Object.Equals()` is `virtual`, so that one uses runtime dispatching based on the type it is called on. So if there is a class that overrides `Equals`, but `==` comparison is used on its instances that is most probably a mistake:
```
class MyClass
{
  //...
  public override bool Equals(object obj)
  {
    // do some checks here instead of a simple "base.Equals(obj);" call
  }
}

class Program
{
  public static void Main()
  {
    var x = new MyClass();
    var y = new MyClass();
    if (x == y) //Noncompliant, didn't we want to call Equals instead? This way we are calling "Object.operator==(object, object)"
    {
      // we never get here as it is a reference comparison
    }
  }
}
```

There is an exception to this rule: if `MyClass` defines `operator ==`, then we can safely assume that the implementer thought about reference and value comparison.
So if `MyClass` looks like this, then we shouldn't report an issue:
```
class MyClass
{
  //...
  public override bool Equals(object obj)
  {
    // do some checks here instead of a simple "base.Equals(obj);" call
  }
  public static bool operator ==(MyClass a, MyClass b)
  {
    //...
  }
  public static bool operator !=(MyClass a, MyClass b)
  {
   //...
  }
}
```

However, when `==` is used on `MyClass`, but it resolves to `object.operator==`, then it can be a bug. This is reported by the compiler (CS0253). This can happen when the types of the expressions on the sides of `==` are not the same. For example `"" == new object()`.

Let's see the same thing with interfaces. Interfaces don't allow defining `operator ==`. 
```
IMyInterface x = GetInterface();
IMyInterface y = GetInterface();

if (x == y) //Noncompliant, if any of the classes implementing IMyInterface overrides Equals. 
{
}
```

So, __to sum up__: we want to report on `a==b`, where, type of `a` or type of `b` overrides `Equals`. If type of `a` or type of `b` is an interface, then we want to check all implementing classes if they override `Equals` or not.

**Comment by tvsonar** [comment]
@dbolkensteyn  could you please double check me on this rule? If you think it's okay, you should assign it to @ganncamp

**Comment by tvsonar** [comment]
RSPEC: https://jira.sonarsource.com/browse/RSPEC-1698
